mypy/plugins/enums.py

```diff
@@ -148,69 +158,87 @@ class SomeEnum:
-                if t is None or not isinstance(t, CallableType)
-            )
-            underlying_type = _first(proper_types)
-            if underlying_type is None:
-                return ctx.default_attr_type
-
-            # At first we try to predict future `value` type if all other items
-            # have the same type. For example, `int`.
-            # If this is the case, we simply return this type.
-            # See https://github.com/python/mypy/pull/9443
-            all_same_value_type = all(
-                proper_type is not None and proper_type == underlying_type
-                for proper_type in proper_types
-            )
-            if all_same_value_type:
-                if underlying_type is not None:
-                    return underlying_type
-
-            # But, after we started treating all `Enum` values as `Final`,
-            # we start to infer types in
-            # `item = 1` as `Literal[1]`, not just `int`.
-            # So, for example types in this `Enum` will all be different:
-            #
-            #  class Ordering(IntEnum):
-            #      one = 1
-            #      two = 2
-            #      three = 3
-            #
-            # We will infer three `Literal` types here.
-            # They are not the same, but they are equivalent.
-            # So, we unify them to make sure `.value` prediction still works.
-            # Result will be `Literal[1] | Literal[2] | Literal[3]` for this case.
-            all_equivalent_types = all(
-                proper_type is not None and is_equivalent(proper_type, underlying_type)
-                for proper_type in proper_types
-            )
-            if all_equivalent_types:
-                return make_simplified_union(cast(Sequence[Type], proper_types))
+            items: List[Union[Instance, LiteralType]] = [ctx.type]
+        elif isinstance(ctx.type, UnionType):
+            tmp_items = [get_proper_type(item) for item in ctx.type.items]
+            items = [item for item in tmp_items if isinstance(item, (Instance, LiteralType))]
+        else:
+            return ctx.default_attr_type
+
+        # As long as mypy doesn't understand attribute creation in __new__,
+        # there is no way to predict the value type if the enum class has a
+        # custom implementation
+        if any(
+            _implements_new(item.type if isinstance(item, Instance) else item.fallback.type)
+            for item in items
+        ):
+            return ctx.default_attr_type
+
+        # there may be duplicate typeinfos in `infos`, but union simplification handles that
+        # stnodes = (info.get(name) for info in infos for name in info.names)
+        stnodes = chain.from_iterable(
+            [item.type.get(name) for name in item.type.names] if isinstance(item, Instance)
+            # this cast is fine because we guard on item.is_enum_literal()
+            else [item.fallback.type.get(cast(str, item.value))]
```

Comment:
Maybe we can restructure to avoid the cast, e.g.
```python
stnodes = chain.from_iterable(
  [item.type.get(name) for name in item.type.names] if isinstance(item, Instance)
  else [item.fallback.type.get(value)] if (value := item.get_enum_value())
  else []
  for item in items
)
```